tools/SampleAnalyzer/Process/JetClustering/JetClusterer.cpp

```diff
@@ -366,81 +413,117 @@ MAbool JetClusterer::Execute(SampleFormat& mySample, EventFormat& myEvent)
+                                                          part.mothers()[0]->decay_vertex().Y(),
+                                                          part.mothers()[0]->decay_vertex().Z(),0.0));
+                muon->setClosestApproach(smeared.closest_approach());
+                muon->setMc(&(part));
+                if (part.pdgid()==13) muon->SetCharge(-1);
+                else muon->SetCharge(+1);
+              }
+
+              // Electrons
+              else if (absid==11)
+              {
+                vetos[i]=true;
+
+                // Smearing the electron momentum
+                MCParticleFormat smeared = mySmearer_->Execute(&part, static_cast<MAint32>(absid));
+                if (smeared.pt() <= 1e-10) continue;
+
+                RecLeptonFormat * elec = myEvent.rec()->GetNewElectron();
+                elec->setMomentum(smeared.momentum());
+                elec->setD0(smeared.d0());
+                elec->setDZ(smeared.dz());
+                elec->setD0Approx(smeared.d0_approx());
+                elec->setDZApprox(smeared.dz_approx());
+                elec->setProductionVertex(MALorentzVector(part.mothers()[0]->decay_vertex().X(),
+                                                          part.mothers()[0]->decay_vertex().Y(),
+                                                          part.mothers()[0]->decay_vertex().Z(),0.0));
+                elec->setClosestApproach(smeared.closest_approach());
+                elec->setMc(&(part));
+                if (part.pdgid()==11) elec->SetCharge(-1);
+                else elec->SetCharge(+1);
+              }
+
+              // Photons
+              else if (absid==22)
+              {
+                if (!LOOP->IrrelevantPhoton(&part,mySample))
+                {
+                    vetos[i]=true;
+
+                    // Smearing the photon momentum
+                    MCParticleFormat smeared = mySmearer_->Execute(&part, static_cast<MAint32>(absid));
+                    if (smeared.pt() <= 1e-10) continue;
+
+                    RecPhotonFormat * photon = myEvent.rec()->GetNewPhoton();
+                    photon->setMomentum(smeared.momentum());
+                    photon->setD0(smeared.d0());
+                    photon->setDZ(smeared.dz());
+                    photon->setD0Approx(smeared.d0_approx());
+                    photon->setDZApprox(smeared.dz_approx());
+                    photon->setProductionVertex(MALorentzVector(part.mothers()[0]->decay_vertex().X(),
+                                                                part.mothers()[0]->decay_vertex().Y(),
+                                                                part.mothers()[0]->decay_vertex().Z(),0.0));
+                    photon->setClosestApproach(smeared.closest_approach());
+                    photon->setMc(&(part));
+                }
+              }
+        }
       }
+#ifdef MA5_FASTJET_MODE
```

Comment:
@BFuks 's proposal: In order to avoid possible clash with Delphes one can write a make file which can accept command line flags such as `make --with-fastjet` or `make --with-delphes`. This will allow different flags to be activated during execution. Or similarly a config file can be written and this config file can then create a make file. In either case this requires the entire `make_writer.py` to be changed accordingly.

Reply:
This can be done by modifying `setup.sh` accordingly;
```shell
export WITH_FASTJET="0"
export WITH_DELPHES="0"
export FASTJET_FLAG=" "

for user in "$@"
do
    if [[ $user == "--with-fastjet" ]]; then
        export WITH_FASTJET="1"
    elif [[ $user == "--with-delphes" ]]; then
        export WITH_DELPHES="1"
    elif [[ $user == "-h" || $user == "--help" ]]; then
        echo -e "usage bla bla bla"
        return 0
    fi
done

if [ $WITH_FASTJET -eq "1" ]  && [ $WITH_DELPHES -eq "1" ]
then
    echo -e $RED"ERROR: FastJet and Delphes can not be executed in the same analysis."$NORMAL
    return 1
fi

if [[ $WITH_FASTJET -eq "1" ]]; then
    export FASTJET_FLAG="-DMA5_FASTJET_MODE"
fi
```

then `setup.sh` can be executed via `source setup.sh --with-fastjet` which will enable the flag for `Makefile`. Then the `Makefile` should be modified via

```bash
CXXFLAGS  = -Wall -std=c++11 -O3 -fPIC -I$(MA5_BASE)/tools/
CXXFLAGS +=  -I./ $(FASTJET_FLAG)
```

where if the source is executed via delphes, fastjet flag will be empty.

@BFuks do you agree?

Reply:
I agree with this scheme. 
 - Maybe we could replace "user" by something like "mode" in the script? 
 - In addition, please also include something to do for unsupported options. For instance, if a user runs ```make --with-a-crazy-option```, then the setup script should return an error (together with the help message).

Reply:
So something like this:

`setup.sh`
```shell
export WITH_FASTJET="0"
export WITH_DELPHES="0"
export FASTJET_FLAG=" "

function usage() {
    echo -e "Usage: source setup.sh [options]"
    echo -e "   -h OR --help   : Prints this very useful text."
    echo -e "   --with-fastjet : Enables the usage of FastJet interface within the analysis."
    echo -e "   --with-delphes : Enables the usage of Delphes interface within the analysis."
}

for user in "$@"
do
    if [ $user == "--with-fastjet" ]
    then
        export WITH_FASTJET="1"
    elif [ $user == "--with-delphes" ]
    then
        export WITH_DELPHES="1"
    elif [ $user == "-h" ] || [ $user == "--help" ]
    then
        usage
        return 0
    else
        echo -e $RED"ERROR: Invalid commandline option."$NORMAL
        usage
        return 1
    fi

done

if [ $WITH_FASTJET -eq "1" ]  && [ $WITH_DELPHES -eq "1" ]
then
    echo -e $RED"ERROR: FastJet and Delphes can not be executed in the same analysis."$NORMAL
    return 1
fi

if [[ $WITH_FASTJET -eq "1" ]]; then
    export FASTJET_FLAG="-DMA5_FASTJET_MODE"
fi
```

usage:
```bash
$ source setup.sh --with-fastjet
$ make -j4
```

Reply:
That looks very good!

Reply:
will update the code accordingly after you are done with your review.